include/sparrow/mp_utils.hpp

```diff
@@ -14,34 +14,257 @@
+
+    template< class L, template<class...> class U >
+    struct is_type_instance_of : std::false_type {};
+
+    template< template<class...> class L, template<class...> class U, class... T>
+        requires std::is_same_v< L<T...>, U<T...> >
+    struct is_type_instance_of< L<T...>, U > : std::true_type {};
+
+    /// `true` if `T` is a concrete type template instanciation of `U` which is a type template.
+    /// Example: is_type_instance_of_v< std::vector<int>, std::vector > == true
+    template< class T, template< class... > class U >
+    constexpr bool is_type_instance_of_v = is_type_instance_of<T, U>::value;
+
+    /// Matches any type which is an instance of `typelist`.
+    /// Examples:
+    ///     static_assert(any_typelist< typelist<int, float, std::vector<int> > > == true);
+    ///     static_assert(any_typelist< std::vector<int> > == false);
+    template< typename TList >
+    concept any_typelist = is_type_instance_of_v<TList, typelist>;
+
+    template <class T, bool is_const>
+    struct constify
+        : std::conditional<is_const, const T, T>
+    {
+    };
+
+    // `constify_t` is required since `std::add_const_t<T&>`
+    // is not `const T&` but `T&`.
+    template <class T, bool is_const>
+    using constify_t = typename constify<T, is_const>::type;
+
+    //////////////////////////////////////////////////
+    //// Type-predicates /////////////////////////////
+
+    /// Matches template types which can be used as type-wrappers for evaluation in type-predicates.
+    template< template< class... > class W, class T >
+    concept type_wrapper = std::same_as<W<T>, typelist<T>>
+                        or std::same_as<W<T>, std::type_identity_t<T>>
+                        ;
+
+    /// Matches template types that can be evaluated at compile-time similarly to `std::true/false_type`
+    /// This makes possible to use predicates provided by the standard.
+    template< template<class> class P, class T >
+    concept ct_type_predicate = requires
+    {
+        { P<T>::value } -> std::convertible_to<bool>;
+    };
+
+    /// Matches types whose instance can be called with an object representing a type to evaluate it.
+    /// The object passed to calling the predicate instance is either a typelist with one element or
+    /// `std::type_identity_t<T>`. Basically a value wrapper for representing a type.
+    /// This is useful to detect type predicates which allow being called like normal functions.
+    template< class P, class T >
+    concept callable_type_predicate = std::semiregular<std::decay_t<P>>
+        and (
+                requires(std::decay_t<P> predicate)
+                {
+                    { predicate(typelist<T>{}) } -> std::convertible_to<bool>;
+                }
+            or
+                requires(std::decay_t<P> predicate)
+                {
+                    { predicate(std::type_identity_t<T>{}) } -> std::convertible_to<bool>;
```

Comment:
> I don't remember why we need to support both type_identity_t and typelist.

I started with `typelist` because that's what the algorithms in this PR work with as argument, to not deviate too far. The support for `type_identity` you suggested simply allows for code taht dont depend on sparrow to still add new type-predicates as callable objects.

> Would the support of type_identity be enough?

Both are fine to me, or do you see a cost I'm not seeing?

> It looks like predicates will be evaluated over types in a typelist via the evaluate function, resulting in a pattern like (evaluate<T> OP  ... OP init); therefore evaluate could accept type_identity_t only and taht would simplify the code. But as always when I have this feeling, I probably issed something.

No you're right, both typelist and type_identity works here so I support both and other than the potential extension without being dependent on sparrow, it would indeed simplify some of the code.

Not sure what to do here. I have a slight preference toward supporting "anything that can act as a type wrapper" so that's why I switched to using the `mpl::type_wrapper` concept instead of being too specific. The implementations of the algorithms could be rewritten with any type wrappers.

Choices. :)

Reply:
> Both are fine to me, or do you see a cost I'm not seeing?

Not really, it was really for simplifying the code. Also concepts with or clauses don't subsume, but I guess it's not really importnat here since we'll probably won'r refine the `callable_type_predicate` concept.

I would say let's keep it as is for now, and use it. And we'll see later if we can really simplify it (if that's really worth it, and it it does not break any code using it).